Cache device on TensorImpl; clean up TensorImpl constructors.

gchanan opened the pull request:
Summary:
1) We cache device on TensorImpl.  This means we can access the device without a virtual function and allows us to more easily extend TensorImpls (because they don't need to figure out how to store the Device for themselves).

2) Clean up TensorImpl APIs.  We had a constructor that took a TensorTypeId and an allocator and would allocate a Storage based on the recognized types of TensorTypeIds.  Instead, we just have two different constructors: one for types with a storage, one without.

gchanan committed:
Cache device on TensorImpl; clean up TensorImpl constructors.

Summary:
1) We cache device on TensorImpl.  This means we can access the device without a virtual function and allows us to more easily extend TensorImpls (because they don't need to figure out how to store the Device for themselves).

2) Clean up TensorImpl APIs.  We had a constructor that took a TensorTypeId and an allocator and would allocate a Storage based on the recognized types of TensorTypeIds.  Instead, we just have two different constructors: one for types with a storage, one without.

gchanan commented:
benchmark incoming.

gchanan commented:
Actually it looks like the benchmarks are broken for previous versions.  But getting the device is faster!

Old:
```
Benchmark                                 Time           CPU Iterations
BM_TensorImplGetDevice                    3 ns          3 ns  279652190
```

New:
```
Benchmark                                 Time           CPU Iterations
BM_TensorImplGetDevice                    2 ns          2 ns  331110583

```

gchanan committed:
Update device_opt_ in set_data.

gchanan commented:
@pytorchbot rebase this please.

Commit:
Merge remote-tracking branch 'origin/master' into HEAD

dzhulgakov commented on c10/core/TensorImpl.h:
```diff
@@ -1401,6 +1397,17 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
+
+  // INVARIANT: When storage is non-null, this Device must
+  // agree with the type meta in storage.
+  c10::optional<c10::Device> device_opt_;
```
why is it optional even? for any tensor except Undefined one it should initialized. For Undefined we can just populate fake value from DeviceType enum, e.g. ONLY_FOR_TEST.

My worry is that c10::optional creates visibility that it's not always populated, though for well-formed tensor it actually is

gchanan replied:
I considered that but it seems worse.  IMO adding an ONLY_FOR_TEST is first of all misleading because it's not only for test, and it moves the complexity from where it should be (when the device really can be non-existent, here) to where it shouldn't be (like now everywhere else in code, where the device must exist).

IMO the correct solution is to:
1) add the interface as above to get the actual device without allocating something.
2) hook up SparseTensor to 1) so the device is properly set in the constructor
3) assert that only UndefinedTensors don't have a devices in the constructor.

ezyang replied:
The basic question is whether or not `Device` should be default constructible or not. We've gotten away with it not being default constructible, and the type-safe person in me likes it.

gchanan commented on aten/src/ATen/SparseTensorImpl.cpp:
```diff
@@ -31,11 +31,15 @@ namespace {
     , indices_(at::empty({1, 0}, at::initialTensorOptions().device(sparseTensorIdToDeviceType(type_id)).dtype(ScalarType::Long)))
-    , values_(at::empty({0}, at::initialTensorOptions().device(sparseTensorIdToDeviceType(type_id)).dtype(data_type))) {}
+    , values_(at::empty({0}, at::initialTensorOptions().device(sparseTensorIdToDeviceType(type_id)).dtype(data_type))) {
+  // FIXME: we need to wait for data-member initialization to know our device because there isn't a device-dependent way
```
not sure how that helps.  The TensorImpl constructor needs to be called first, but the only way we currently get the actual device is via the Storage, but any Storage isn't created until later.

We should just add a way to get the 'actual' device for any device type without allocating, but that seemed like a big enough change to include in a different diff.

ezyang replied:
Ah, the pain and pleasure of C++ subclass constructor ordering >.>

dzhulgakov replied:
I mean having something like:

```
SparseTensorImpl::SparseTensorImpl(at::TensorTypeId type_id, const caffe2::TypeMeta& data_type)
: SparseTensorImpl(type_id, data_type,
    at::empty({1, 0}, at::initialTensorOptions().device(sparseTensorIdToDeviceType(type_id)).dtype(ScalarType::Long))),
    at::empty({0}, at::initialTensorOptions().device(sparseTensorIdToDeviceType(type_id)).dtype(data_type))) {}

/*private*/ SparseTensorImpl::SparseTensorImpl(at::TensorTypeId type_id, const caffe2::TypeMeta& data_type, Tensor indices, Tensor values) : 
TensorImpl(type_id, data_type, values.device(), false),
...
, indices_(indices),
, values_(values) {}
```

would it work?

gchanan replied:
that might work!  Let me give it a go.

gchanan commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -320,7 +320,7 @@ struct ContainerTensor : public at::TensorImpl {
       : TensorImpl(
```
Sure, I'll do it in another diff.

dzhulgakov replied:
Just kill it here so you don't even need to change it

ezyang commented on c10/core/TensorImpl.h:
```diff
@@ -1475,12 +1482,13 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
 //
 static_assert(sizeof(void*) != sizeof(int64_t) || // if 64-bit...
-              sizeof(TensorImpl) == sizeof(int64_t) * 26,
+              sizeof(TensorImpl) == sizeof(int64_t) * 27,
```
I'm... actually kind of impressed this only took one extra word

ezyang commented on c10/core/TensorImpl.h:
```diff
@@ -891,15 +879,18 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
    * The device of a Tensor; e.g., Device(kCUDA, 1) (the 1-index CUDA
    * device).
    */
   Device GetDevice() const {
```
Bloop, a bootcamper removed this function but they never landed it XD

ezyang commented on c10/core/TensorImpl.h:
```diff
@@ -891,15 +879,18 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
    */
   DeviceType device_type() const {
```
Why don't you call `device()` here? You'd get a better error message in that case, no?

gchanan replied:
not a bad idea...I was looking into if we could just kill this (TensorImpl should be minimal, this doesn't add any functionality), but didn't follow through.

ezyang commented on c10/core/TensorImpl.h:
```diff
@@ -361,31 +359,25 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
   }
 
   int64_t get_device() const {
```
We should kill this entirely at some point. Not this PR though.

gchanan replied:
do you mean to only have `device()` in TensorImpl?

ezyang reviewed (approved):
Sparse is a little goofy but fine to unblock

gchanan committed:
Materialize a non-default device for C2 legacy storage.

It's not intended that Storages have 'default' CUDA devices, but this is allowable via the Storage::create_legacy codepath.

This also messages with device_caching, because the initial cache is obtained from the Storage, which may have a 'default' device.

Instead, we materialize a device by allocating 0 bytes via the allocator.

gchanan commented:
@pytorchbot rebase this please.

Commit:
Merge remote-tracking branch 'origin/master' into HEAD

gchanan commented:
I incorporated https://github.com/pytorch/pytorch/pull/18605 into here.  I'm not sure if it's correct, but I didn't check if it's possible to use ghstack to stack this on top of a new PR, so I just threw it in here to get test signal.

dzhulgakov commented on c10/core/TensorImpl.h:
```diff
@@ -1184,6 +1178,7 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
         }
       }
       data_type_ = meta;
+      // NB: device is not changed
```
I'd rather not do https://github.com/pytorch/pytorch/pull/18605 and instead reset the device - around line 1215 below. Non-initialized tensors in C2 can be created at random places (e.g. in static fields) but the actual data pointer would be populated only at raw_mutable_data() call and the device may have changed in the mean time. It's a weird behavior, but still...

gchanan committed:
Kill ContainerTensor.

gchanan committed:
Proxy to SparseTensorImpl constructor in order to initialize device in the correct order.

gchanan committed:
Revert "Materialize a non-default device for C2 legacy storage."

This reverts commit 0bcd785606a0ff9ee233be7dec42a3d1b7edebc6.

gchanan committed:
Update device_opt_ in raw_mutable_data.

gchanan committed:
Add assert that for all 'real' tensors, device is really set.

dzhulgakov commented on aten/src/ATen/SparseTensorImpl.h:
```diff
@@ -210,11 +210,8 @@ struct CAFFE2_API SparseTensorImpl : public TensorImpl {
-  }
-
+private:
+    explicit SparseTensorImpl(at::TensorTypeId, const caffe2::TypeMeta&, at::Tensor indices, at::Tensor values);
```
nit: no need for explicit

gchanan replied:
the other constructor had it, but ya, I should have removed both.

dzhulgakov commented on c10/core/TensorImpl.h:
```diff
@@ -1219,6 +1214,7 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
       }
       storage_.set_numel(numel_);
       AT_ASSERT(storage_offset_ == 0); // because we just reallocated
+      device_opt_ = storage_.device();
```
I'd add a comment:
```
// raw_mutable_data might allocate or reallocate memory on a different device (e.g. in CUDA case) - we need to update the cache on the Tensor side
```

dzhulgakov reviewed (approved):
Thanks for updating! Looks good to me!

gchanan committed:
Update xla_tensor_test to not use a Storage.

gchanan committed:
Loosen assert constraints to allow default devices.

gchanan commented:
@pytorchbot rebase this please.

Commit:
Merge remote-tracking branch 'origin/master' into HEAD

gchanan committed:
Correctly update cached device in SparseTensorImpl shallow copy.

gchanan commented:
closing in favor of https://github.com/pytorch/pytorch/pull/18833.
